packages/loader/container-loader/src/retriableDocumentStorageService.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { CreateContainerError } from "@fluidframework/container-utils";
+import { IDocumentStorageService } from "@fluidframework/driver-definitions";
+import { canRetryOnError, DocumentStorageServiceProxy } from "@fluidframework/driver-utils";
+import { ISnapshotTree, IVersion } from "@fluidframework/protocol-definitions";
+import { Container, RetryFor } from "./container";
+import { getRetryDelayFromError } from "./deltaManager";
+
+export class RetriableDocumentStorageService extends DocumentStorageServiceProxy {
+    private static callsWaiting: number = 0;
+    private static futureTimeTillWait: number = 0;
+    private disposed = false;
+    constructor(
+        internalStorageService: IDocumentStorageService,
+        private readonly container: Pick<Container, "emitDelayInfo" | "cancelDelayInfo">,
+    ) {
+        super(internalStorageService);
+    }
+
+    public dispose() {
+        this.disposed = true;
+    }
+
+    public async getSnapshotTree(version?: IVersion): Promise<ISnapshotTree | null> {
+        return this.readWithRetry(async () => this.internalStorageService.getSnapshotTree(version));
+    }
+
+    public async read(blobId: string): Promise<string> {
+        return this.readWithRetry(async () => this.internalStorageService.read(blobId));
+    }
+
+    public async readBlob(id: string): Promise<ArrayBufferLike> {
+        return this.readWithRetry(async () => this.internalStorageService.readBlob(id));
+    }
+
+    public async readString(id: string): Promise<string> {
+        return this.readWithRetry(async () => this.internalStorageService.readString(id));
+    }
+
+    private async delay(timeMs: number): Promise<void> {
+        return new Promise((resolve) => setTimeout(() => resolve(), timeMs));
+    }
+
+    private async readWithRetry<T>(api: () => Promise<T>): Promise<T> {
+        let result: T | undefined;
+        let success: boolean | undefined;
+        let retryAfter = 0;
+        do {
+            try {
+                result = await api();
+                if (success === false) {
+                    RetriableDocumentStorageService.callsWaiting -= 1;
+                    if (RetriableDocumentStorageService.callsWaiting === 0) {
+                        RetriableDocumentStorageService.futureTimeTillWait = 0;
+                        this.container.cancelDelayInfo(RetryFor.Storage);
+                    }
+                }
+                success = true;
```

Comment:
you can return result from here, and thus make success a true boolean (not tri-state) - code would be more readable.